### DISCUSSION TITLE
How to handle multi-languages language servers

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1222] [Q&A] [answered]
Hi there,

In our project, we have different languages/diagrams have the same document, so we overridden `ServiceRegistry` and register each language and it services and used a detector to check what language it is.
But it appears that the language server doesn't work anymore, shouldn't it switch between language servers when it identifies the keyword? It does that for the parser and other services.

- Using the diagram service itself:

    ![image](https://github.com/eclipse-langium/langium/assets/87622592/85cc0c4b-aa00-4511-ba54-6404101b7818)

- Using the wrapper service that contains all diagrams:

    ![image](https://github.com/eclipse-langium/langium/assets/87622592/aaf2f4b5-dfb4-498a-b703-0010cc2395f8)

---

Is this an issue? Or we're messing something? Should I create a clone of the project?

**Comment by msujew** [comment]
Do you get any error from the language server when you request completions? Does it generate diagnostics as expected? The first issue to rule out would be to identify whether the language service selector/registry is at fault or whether the logic doesn't play nicely with the completion provider.

**Reply by Yokozuna59** [reply]
> Do you get any error from the language server when you request completions?

No.

![image](https://github.com/eclipse-langium/langium/assets/87622592/0254f0cb-5c5f-4a41-a67a-8b145ba1d948)

> Does it generate diagnostics as expected?

I'm not really sure how to check that.

**Reply by msujew** [reply]
Just type in something invalid, such as `info ßßß` (hopefully that is actually invalid data). It should complain about a syntax error.

**Reply by Yokozuna59** [reply]
![image](https://github.com/eclipse-langium/langium/assets/87622592/c76f1ee8-6c2b-4b69-830d-3d2e1eb7617c)

Other services works fine expect completions of the language server.

**Reply by msujew** [reply]
Ok that seems to rule out the issue that it can't find the info service. Do you have a reproducible example I can take a look at? I'm curious what goes wrong in the completion provider.

**Reply by Yokozuna59** [reply]
Sure, here is it: https://github.com/Yokozuna59/multi-lang-example. It took me more time to add related files only.

To build the playground:

```bash
pnpm --filter playground build
```


And to start the editor:

```bash
pnpm --filter playground dev
```

**Comment by Yokozuna59** [comment]
The weird part is that when running the languages as a VSCode extension, it doesn't run any services, not even the parser.
But if I used the individual language service, it would only enable the parser when the document is saved. It doesn't show real-time error highlighting or autocomplete for keywords.

**Reply by msujew** [reply]
Can you share a branch/fork of the project as a VSCode extension? I'm having a bit of trouble building/debugging the repo you linked.

**Reply by Yokozuna59** [reply]
@msujew Done, pushed to master.

> I'm having a bit of trouble building/debugging the repo you linked.

Can you please share them?

**Reply by msujew** [reply]
Sure, seems like the import [here](https://github.com/Yokozuna59/multi-lang-example/blob/d71f444e83c2e91ecf82c77d6ddba0687cf1b5eb/packages/lsp/src/index.ts#L1) doesn't exist.

**Reply by msujew** [reply]
The vscode extension doesn't work due to an id mismatch, see [here](https://github.com/Yokozuna59/multi-lang-example/blob/d71f444e83c2e91ecf82c77d6ddba0687cf1b5eb/packages/mermaid-vscode/package.json#L47).

**Reply by msujew** [reply, marked as answer]
Your service registry is missing an override for the `all` getter:

```ts
override get all(): LangiumServices[] {
    return Object.values(this.services);
}
```

The language server tries to identify whether it needs to register a completion provider based on the info that any service has a completion provider at all. It does this by looking through the `all` property. In your case, it always stays empty, so no language services are registered.

I've investigated the completion issue a bit and saw that the completion parser arrives at an unexpected state:

![image](https://github.com/eclipse-langium/langium/assets/4377073/d774038b-21e5-4f01-a5b0-20d664e6311c)

The added `\n` token doesn't appear in the input, but gets added by the lexer. As mentioned before, completion in whitespace-sensitive languages doesn't work well currently. PRs improving this are appreciated though.

**Reply by Yokozuna59** [reply]
> Sure, seems like the import [here](https://github.com/Yokozuna59/multi-lang-example/blob/d71f444e83c2e91ecf82c77d6ddba0687cf1b5eb/packages/lsp/src/index.ts#L1) doesn't exist.

I've it locally, it was for overriding services to add LSP services, but I didn't use it yet so I didn't push it.

> The vscode extension doesn't work due to an id mismatch, see [here](https://github.com/Yokozuna59/multi-lang-example/blob/d71f444e83c2e91ecf82c77d6ddba0687cf1b5eb/packages/mermaid-vscode/package.json#L47).

The id is indented, I used this id to not interfere with existing extensions that uses this id.
But I guess I didn't change these:

https://github.com/Yokozuna59/multi-lang-example/blob/d71f444e83c2e91ecf82c77d6ddba0687cf1b5eb/packages/mermaid-vscode/src/extension.ts#L48-L57

**Reply by Yokozuna59** [reply]
> Your service registry is missing an override for the `all` getter:
> 
> ```ts
> override get all(): LangiumServices[] {
>     return Object.values(this.services);
> }
> ```
> 
> The language server tries to identify whether it needs to register a completion provider based on the info that any service has a completion provider at all. It does this by looking through the `all` property. In your case, it always stays empty, so no language services are registered.

I see, thanks! It's working as expected. Except `showData` from `pie`, maybe it's because of the whitespace-sensitive as you stated.

> I've investigated the completion issue a bit and saw that the completion parser arrives at an unexpected state:
> 
> The added `\n` token doesn't appear in the input, but gets added by the lexer. As mentioned before, completion in whitespace-sensitive languages doesn't work well currently. PRs improving this are appreciated though.

Oh, I see. I don't really know how to resolve such error. Would using `EOF` would resolve such error?

**Reply by Yokozuna59** [reply]
The completions is currently working in the vscode extension, but the parser doesn't.
And it keeps throwing error:

```
Error:  RangeError: Maximum call stack size exceeded
    at Object.get (packages/mermaid-vscode/dist/main.cjs:65:57744)
    at e8.documents (packages/mermaid-vscode/dist/main.cjs:203:1526)
    at e8.getServices (packages/mermaid-vscode/dist/main.cjs:203:1722)
    at packages/mermaid-vscode/dist/main.cjs:88:20796
    at get textDocument [as textDocument] (packages/mermaid-vscode/dist/main.cjs:88:20164)
    at e8.getServices (packages/mermaid-vscode/dist/main.cjs:203:1848)
    at packages/mermaid-vscode/dist/main.cjs:88:20796
    at get textDocument [as textDocument] (packages/mermaid-vscode/dist/main.cjs:88:20164)
    at e8.getServices (packages/mermaid-vscode/dist/main.cjs:203:1848)
    at packages/mermaid-vscode/dist/main.cjs:88:20796
```

The browser works fine though.

**Reply by msujew** [reply]
I see. You probably need to override the [`createTextDocumentGetter`](https://github.com/eclipse-langium/langium/blob/a4c7a58e9a79199d0f69cd3a9479c6ddfe99de21/packages/langium/src/workspace/documents.ts#L257-L265) method to get the type from the `text` instead of using the service registry.

**Reply by Yokozuna59** [reply]
Oh, okay. Can you please share the new expected flow or code? I modified both `MermaidServiceRegistry` and `MermaidDocumentFactory`, but I think I'm missing something because it's not working anymore.

**Reply by Yokozuna59** [reply]
> I see. You probably need to override the [`createTextDocumentGetter`](https://github.com/eclipse-langium/langium/blob/a4c7a58e9a79199d0f69cd3a9479c6ddfe99de21/packages/langium/src/workspace/documents.ts#L257-L265) method to get the type from the `text` instead of using the service registry.

I'm not really sure about what approach you're suggesting, but the following one works fine for me:

```ts
export class MermaidDocumentsFactory extends DefaultLangiumDocumentFactory {
  private documents: {
    uri: URI;
    languageId: string;
    textDocument: TextDocument;
  }[] = [];

  protected override parse<T extends AstNode>(
    uri: URI,
    text: string,
  ): ParseResult<T> {
    // ...
  }

  protected override createTextDocumentGetter(
    uri: URI,
    text?: string,
  ): () => TextDocument {
    const { languageId } =
      this.serviceRegistry.getServices(uri).LanguageMetaData;

    return () => {
      return (
        this.textDocumentsArray.find((textDocument) => {
          return (
            textDocument.languageId === languageId &&
            textDocument.uri === uri.toString()
          );
        }) ?? this.createNewTextDocument(uri, languageId, text)
      );
    };
  }

  private createNewTextDocument(
    uri: URI,
    languageId: string,
    text?: string,
  ): TextDocument {
    const textDocument = TextDocument.create(
      uri.toString(),
      languageId,
      0,
      text ?? '',
    );
    this.textDocumentsArray.push(textDocument);
    return textDocument;
  }
}
```
